ruby_event_store_rom_sql/lib/ruby_event_store_rom_sql/event_repository.rb

```diff
@@ -0,0 +1,151 @@
+      @event_streams = ROM::Repositories::EventStreams.new(rom)
+      @mapper        = mapper
+    end
+
+    def append_to_stream(events, stream_name, expected_version)
+      add_to_stream(events, stream_name, expected_version, true) do |event|
+        @events.create(map_to_serialized_record(event))
+        event.event_id
+      end
+    end
+
+    def link_to_stream(event_ids, stream_name, expected_version)
+      @events.detect_invalid_event_ids(normalize_to_array(event_ids)).each do |id|
+        raise RubyEventStore::EventNotFound.new(id)
+      end
+      add_to_stream(normalize_to_array(event_ids), stream_name, expected_version, nil) do |event_id|
+        event_id
+      end
+    end
+
+    def delete_stream(stream_name)
+      @event_streams.delete_events_for(stream_name)
+    end
+
+    def has_event?(event_id)
+      @events.has_event?(event_id)
+    end
+
+    def last_stream_event(stream_name)
+      map_to_event @events.last_stream_event(stream_name)
+    end
+
+    def read_events_forward(stream_name, after_event_id, count)
+       @events.read_events_forward(stream_name, after: after_event_id, limit: count).map(&method(:map_to_event))
+    end
+
+    def read_events_backward(stream_name, before_event_id, count)
+      @events.read_events_backward(stream_name, before: before_event_id, limit: count).map(&method(:map_to_event))
+    end
+
+    def read_stream_events_forward(stream_name)
+      @events.read_events_forward(stream_name).map(&method(:map_to_event))
+    end
+
+    def read_stream_events_backward(stream_name)
+      @events.read_events_backward(stream_name).map(&method(:map_to_event))
+    end
+
+    def read_all_streams_forward(after_event_id, count)
+      read_events_forward(RubyEventStore::GLOBAL_STREAM, after_event_id, count)
+    end
+
+    def read_all_streams_backward(before_event_id, count)
+      read_events_backward(RubyEventStore::GLOBAL_STREAM, before_event_id, count)
+    end
+
+    def read_event(event_id)
+      map_to_event @events.fetch(event_id)
+    end
+
+    def get_all_streams
+      @event_streams.get_all_streams
+    end
+
+    def add_metadata(event, key, value)
+      @mapper.add_metadata(event, key, value)
+    end
+
+    private
+
+    def add_to_stream(event_ids, stream_name, expected_version, include_global, &to_event_id)
+      raise RubyEventStore::InvalidExpectedVersion if !expected_version.equal?(:any) &&
+                                                      stream_name.eql?(RubyEventStore::GLOBAL_STREAM)
+
+      event_ids = normalize_to_array(event_ids)
+      expected_version = normalize_expected_version(expected_version, stream_name)
+
+      # TODO: Move this into repository (don't expose db internals)!!
+      @event_streams.event_streams.transaction(savepoint: true) do
+        in_stream = event_ids.flat_map.with_index do |event_id, index|
+          position = compute_position(expected_version, index)
+          event_id = to_event_id.call(event_id)
```

Comment:
@pawelpacana @paneq this logic is taken from the AR version, where the events are persisted in the loop and the `event_id` returned. Is there any reason that the events aren't bulk-inserted into the database in one step, and then linking them to the stream in a separate step? Is it simply to reuse the same loop? (I don't know how often there's the scenario of adding more than a few events at a time, but separating the steps would make things a little clearer.

Reply:
> Is there any reason that the events aren't bulk-inserted into the database in one step

that would make sense. Historically it was only possible to store a single event. Later we added a possibly to handle many of them. The code might be this way due to this historical reasons only,

Reply:
@paneq refactored ROM to do things in batches.